geolocatezip.3
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man 5.01 (Pod::Simple 3.43)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\" ========================================================================
.\"
.IX Title "GEOLOCATEZIP 1"
.TH GEOLOCATEZIP 1 2019-11-12 "perl v5.38.2" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
geolocatezip.pl
.SH SYNOPSIS
.IX Header "SYNOPSIS"
\&./geolocatezip.pl, by (c) Jim Turner
.SH DESCRIPTION
.IX Header "DESCRIPTION"
geolocatezip.pl attempts to first fetch the users current IP address, then uses that 
to fetch the zipcode for that IP address using geolocation.  Failing that, it attempts 
to fetch the city and state for the IP address and then uses a postal service site 
to attempt to look up a zip-code for that city and state.  If multiple zip-codes are 
returned, then the last one is returned.  The zip-code found is written to STDOUT.
.PP
The purpose of this program is for use by TkWeather in order to get the weather 
hopefully for the location the user's computer is actually in.  This is useful if 
the user is a frequent traveler and using a laptop.
.PP
NOTE:  works for U.S., maybe Canada too; for now.
.SH "LICENSE AND COPYRIGHT"
.IX Header "LICENSE AND COPYRIGHT"
Copyright 2015\-2019 Jim Turner.
.PP
This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:
.PP
<http://www.perlfoundation.org/artistic_license_2_0>
.PP
Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.
.PP
If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.
.PP
This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.
.PP
This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.
.PP
Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.SS OBSOLETE
.IX Subsection "OBSOLETE"
if ($html =~ m#(?:Region|State)(?:\e<\e/\ew+\e>)+\es+(\ew+)#i) {  #EXTRACT THE STATE (SITE CURRENTLY RETURNS A FULL STATE NAME):
	\f(CW$state\fR = \f(CW$1\fR;
print "\-\-\-st=$state=$2=\en";
	my \f(CW%stateHash\fR = (
			'TEXAS' => 'TX',
			'ALABAMA' => 'AL',
			'ALASKA' => 'AK',
			'ALBERTA' => 'AB',
			'ARIZONA' => 'AZ',
			'ARKANSAS' => 'AR',
			'BRITISH COLUMBIA' => 'BC',
			'CALIFORNIA' => 'CA',
			'COLORADO' => 'CO',
			'CONNECTICUT' => 'CT',
			'DELAWARE' => 'DE',
			'DISTRICT OF COLUMBIA' => 'DC',
			'FLORIDA' => 'FL',
			'GEORGIA' => 'GA',
			'HAWAII' => 'HI',
			'IDAHO' => 'ID',
			'ILLINOIS' => 'IL',
			'INDIANA' => 'IN',
			'IOWA' => 'IA',
			'KANSAS' => 'KS',
			'KENTUCKY' => 'KY',
			'LOUISIANA' => 'LA',
			'MAINE' => 'ME',
			'MANITOBA' => 'MB',
			'MARYLAND' => 'MD',
			'MASSACHUSETTS' => 'MA',
			'MICHIGAN' => 'MI',
			'MINNESOTA' => 'MN',
			'MISSISSIPPI' => 'MS',
			'MISSOURI' => 'MO',
			'MONTANA' => 'MT',
			'NEBRASKA' => 'NE',
			'NEVADA' => 'NV',
			'NEW BRUNSWICK' => 'NB',
			'NEW HAMPSHIRE' => 'NH',
			'NEW JERSEY' => 'NJ',
			'NEW MEXICO' => 'NM',
			'NEW YORK' => 'NY',
			'NEWFOUNDLAND AND LABRADOR' => 'NF',
			'NORTH CAROLINA' => 'NC',
			'NORTH DAKOTA' => 'ND',
			'NORTHWEST TERRITORIES' => 'NT',
			'NOVA SCOTIA' => 'NS',
			'OHIO' => 'OH',
			'OKLAHOMA' => 'OK',
			'ONTARIO' => 'ON',
			'OREGON' => 'OR',
			'PENNSYLVANIA' => 'PA',
			'PRINCE EDWARD ISLAND' => 'PE',
			'QUEBEC' => 'QC',
			'RHODE ISLAND' => 'RI',
			'SASKATCHEWAN' => 'SK',
			'SOUTH CAROLINA' => 'SC',
			'SOUTH DAKOTA' => 'SD',
			'TENNESSEE' => 'TN',
			'UTAH' => 'UT',
			'VERMONT' => 'VT',
			'VIRGINIA' => 'VA',
			'WASHINGTON' => 'WA',
			'WEST VIRGINIA' => 'WV',
			'WISCONSIN' => 'WI',
			'WYOMING' => 'WY',
			'YUKON' => 'YT'
	);
	\f(CW$state\fR =~ tr/a\-z/A\-Z/;
	\f(CW$state\fR = \f(CW$stateHash\fR{$state}  if (defined \f(CW$stateHash\fR{$state});  #CONVERT STATE NAMES TO PROPER CODE:
}
if ($html =~ m#City(?:\e<\e/\ew+\e>)+\es+(\ew+)#i) {   #EXTRACT THE CITY:
	\f(CW$city\fR = \f(CW$1\fR;
}
.PP
print "\-city=$city= st=$state=\en";
if ($city && \f(CW$state\fR) {    #WE HAVE A VALID CITY, STATE \- NOW TRY TO FETCH A ZIP-CODE:
	\f(CW$zipSite\fR =~ s/\e<city\e>/$city/g;
	\f(CW$zipSite\fR =~ s/\e<state>/$state/g;
#	\f(CW$html\fR = LWP::Simple::get($zipSite);
my \f(CW$lwpcurl\fR = LWP::Curl\->new(timeout => 10);
die "f:Could not create curl object?!"  unless ($lwpcurl);
print "\-\-ZIPSITE=$zipSite=\en";
\&\f(CW$html\fR = \f(CW$lwpcurl\fR\->get($zipSite);  #STEP 1:  FETCH IP AND IT'S CITY, STATE; OR ZIP:
print "\-\-\-html=".substr($html,0,40)."=\en";
	while ($html =~ s!\e<td valign\e=\e"top\e" class\e=\e"main\e" style\e=\e"background\e:url\e(images\e/table_gray.gif\e); padding\e:5px 10px\e;\e" headers\e=\e"units\e"\e>\es*(\ed+)!!s) {
		\f(CW$zip\fR = \f(CW$1\fR;
print "\-\-\-\-\-\-\-zip=$zip=\en";
	}
	print \f(CW$zip\fR;
	\fBexit\fR\|(0);
}


Changes
v1.05:	Eliminated "Geo::Weather" module requirement - www.w3.weather.com 
changed their HTML *AGAIN* and Geo::Weather is apparently keeping up, so I 
now scrape their page myself.
Added "browser" and "weatherurl" options, see manpage.

v1.06: Not released.

v1.07: Added ability to fetch weather from wunderground.
v1.09: Fixed install issues
v1.31: Added www.noaa.gov as a 3rd datasource choice.  Added metric option.
       Added warning / alert color to background.
v1.33: Fix for slight changes in Wunderground site (icon, wind, and visibility).
v1.34: -not released-
v1.35: Fix for slight changes in Wunderground & NOAA sites.
v1.37: Fix to accomodate alternate url in NOAA for the icon image.
v1.8:  Add fourth weather site option (Weatherbug).
v1.83: Create Windows version, Windows self-extracting setup and unify 
       with Linux version.
v1.84: Minor font change, fix Windows installer bug.
v1.87: Minor changes to accomodate weather site changes.
v2.1:  Added geolocation option for automatically fetching the weather for 
       for the zip-code in/near your ISP.
v2.21: Fixed weathersite changes
v2.22: Fixed wunderground icon url chg.
v2.32: Fix for massive changes to weatherbug's site - also added check so that 
       if lwp.simple.get fails to fetch site, we now try wget.  This needed for 
       weatherbug's "mobile" (alternate) site.  Also made slight tweaks to the 
       "misery index" calculation to factor both rel. humidity & dewpoint.
v2.33: Workaround for foobarred lwp-request (on my box) that now refuses to 
	  download images anymore by failing over to "wget".  Also tweaks for 
	  weather.com site changes. 
v2.60: Removed weatherbug from the choices of sites - they've become too spammed 
       up and it is no longer possible to scrape off the needed information.
v2.61: Removed weather.com from choices - too javascripted up, data no longer scrapeable - grrrrrr!
v2.71: Maintenance release to fix up changes in the sites.
       Added a "hot warning color" (hwarningcolor).
       Fixed issue w/geolocatezip.pl due to changes in the USPS website.
v2.72: Fix out-of-bounds weather-site array index.
v2.73: Reset image to default "smiling sun" image if TkWeather becomes unable 
       to pull (update) from any site.  Also, don't recalculate warning 
       background colors if no valid data was pulled.
v2.8:  Redesign regices for Wunderground site (they changed it up a bunch).
       Update license & copyright notice.
------- LONG DORMANT PERIOD -------
v3.4:  * NOAA is now the only weather site supported, as during the intrim, 
       Weather.com, Wunderground and Weatherbug (all commercial sites) have 
       all completely ofuscated their sites to where it's no longer possible to 
       scrape both an icon and the pertinent weather info text.
       * Numerous other updates, fixes & tweaks made internally.
       * "Misery Index" uses an improved formula & is scaled to 100 instead of 
       150 to look/feel more like a real feirenheight temperature value.
       * geolocatezip.pl has not changed.
       * Also included my tkweather.cfg (zip-code changed for privacy) for a 
       working example.
v3.41: * Fix failure to fully flash-cycle through the displays back to the 
       current one when in weather-alert status.
       * Change default re-fetch time from 1 to 3 minutes after a failure to 
       retrieve html from the site.
v3.42: * Fix ignoring of cycle-time between weather updates (checkmsec).
v3.43: * Fix 2nd icon (forcast) image not always working due to slight changes 
       to NOAA's site.
       * Regenerate POD docs using latest (v5.x) vsn. of pod2man.


geolocatezip.pl
#!/usr/bin/perl

=head1 NAME

geolocatezip.pl

=head1 SYNOPSIS

./geolocatezip.pl, by (c) Jim Turner

=head1 DESCRIPTION

geolocatezip.pl attempts to first fetch the users current IP address, then uses that 
to fetch the zipcode for that IP address using geolocation.  Failing that, it attempts 
to fetch the city and state for the IP address and then uses a postal service site 
to attempt to look up a zip-code for that city and state.  If multiple zip-codes are 
returned, then the last one is returned.  The zip-code found is written to STDOUT.

The purpose of this program is for use by TkWeather in order to get the weather 
hopefully for the location the user's computer is actually in.  This is useful if 
the user is a frequent traveler and using a laptop.

NOTE:  works for U.S., maybe Canada too; for now.

=head1 LICENSE AND COPYRIGHT

Copyright 2015-2019 Jim Turner.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

#use LWP::Simple;
my $haveCurl = 0;
eval "use LWP::Curl; \$haveCurl = 1; 1";

#SITE FOR LOOKING UP USER'S TRUE IP-ADDRESS:
my $ipSite = 'https://www.ipinfodb.com/';
#my $html = LWP::Simple::get($ipSite);  #STEP 1:  FETCH IP AND IT'S CITY, STATE; OR ZIP:
my $html = '';

#STEP 1:  FETCH IP AND IT'S CITY, STATE; OR ZIP:
if ($haveCurl) {
	my $lwpcurl = LWP::Curl->new(timeout => 10);
	if ($lwpcurl) {
		$html = $lwpcurl->get($ipSite);
	} else {
		warn "f:Could not create curl object?, trying wget!...";
	}
}
$html = `wget -t 2 -T 20 -O- -o /dev/null \"$ipSite\" 2>/dev/null `  unless ($html);

if ($html =~ /Postcode\D+(\d+)/iso) {
	$zip = $1;
#print "-zip=$zip=\n";
	if ($zip =~ /\d\d\d\d\d/o) {   #IP INFO INCLUDES A VALID ZIP, RETURN THAT; WE'RE DONE!
		print $zip;
		exit(0);
	}
}

=head2 OBSOLETE

if ($html =~ m#(?:Region|State)(?:\<\/\w+\>)+\s+(\w+)#i) {  #EXTRACT THE STATE (SITE CURRENTLY RETURNS A FULL STATE NAME):
	$state = $1;
print "---st=$state=$2=\n";
	my %stateHash = (
			'TEXAS' => 'TX',
			'ALABAMA' => 'AL',
			'ALASKA' => 'AK',
			'ALBERTA' => 'AB',
			'ARIZONA' => 'AZ',
			'ARKANSAS' => 'AR',
			'BRITISH COLUMBIA' => 'BC',
			'CALIFORNIA' => 'CA',
			'COLORADO' => 'CO',
			'CONNECTICUT' => 'CT',
			'DELAWARE' => 'DE',
			'DISTRICT OF COLUMBIA' => 'DC',
			'FLORIDA' => 'FL',
			'GEORGIA' => 'GA',
			'HAWAII' => 'HI',
			'IDAHO' => 'ID',
			'ILLINOIS' => 'IL',
			'INDIANA' => 'IN',
			'IOWA' => 'IA',
			'KANSAS' => 'KS',
			'KENTUCKY' => 'KY',
			'LOUISIANA' => 'LA',
			'MAINE' => 'ME',
			'MANITOBA' => 'MB',
			'MARYLAND' => 'MD',
			'MASSACHUSETTS' => 'MA',
			'MICHIGAN' => 'MI',
			'MINNESOTA' => 'MN',
			'MISSISSIPPI' => 'MS',
			'MISSOURI' => 'MO',
			'MONTANA' => 'MT',
			'NEBRASKA' => 'NE',
			'NEVADA' => 'NV',
			'NEW BRUNSWICK' => 'NB',
			'NEW HAMPSHIRE' => 'NH',
			'NEW JERSEY' => 'NJ',
			'NEW MEXICO' => 'NM',
			'NEW YORK' => 'NY',
			'NEWFOUNDLAND AND LABRADOR' => 'NF',
			'NORTH CAROLINA' => 'NC',
			'NORTH DAKOTA' => 'ND',
			'NORTHWEST TERRITORIES' => 'NT',
			'NOVA SCOTIA' => 'NS',
			'OHIO' => 'OH',
			'OKLAHOMA' => 'OK',
			'ONTARIO' => 'ON',
			'OREGON' => 'OR',
			'PENNSYLVANIA' => 'PA',
			'PRINCE EDWARD ISLAND' => 'PE',
			'QUEBEC' => 'QC',
			'RHODE ISLAND' => 'RI',
			'SASKATCHEWAN' => 'SK',
			'SOUTH CAROLINA' => 'SC',
			'SOUTH DAKOTA' => 'SD',
			'TENNESSEE' => 'TN',
			'UTAH' => 'UT',
			'VERMONT' => 'VT',
			'VIRGINIA' => 'VA',
			'WASHINGTON' => 'WA',
			'WEST VIRGINIA' => 'WV',
			'WISCONSIN' => 'WI',
			'WYOMING' => 'WY',
			'YUKON' => 'YT'
	);
	$state =~ tr/a-z/A-Z/;
	$state = $stateHash{$state}  if (defined $stateHash{$state});  #CONVERT STATE NAMES TO PROPER CODE:
}
if ($html =~ m#City(?:\<\/\w+\>)+\s+(\w+)#i) {   #EXTRACT THE CITY:
	$city = $1;
}

print "-city=$city= st=$state=\n";
if ($city && $state) {    #WE HAVE A VALID CITY, STATE - NOW TRY TO FETCH A ZIP-CODE:
	$zipSite =~ s/\<city\>/$city/g;
	$zipSite =~ s/\<state>/$state/g;
#	$html = LWP::Simple::get($zipSite);
my $lwpcurl = LWP::Curl->new(timeout => 10);
die "f:Could not create curl object?!"  unless ($lwpcurl);
print "--ZIPSITE=$zipSite=\n";
$html = $lwpcurl->get($zipSite);  #STEP 1:  FETCH IP AND IT'S CITY, STATE; OR ZIP:
print "---html=".substr($html,0,40)."=\n";
	while ($html =~ s!\<td valign\=\"top\" class\=\"main\" style\=\"background\:url\(images\/table_gray.gif\); padding\:5px 10px\;\" headers\=\"units\"\>\s*(\d+)!!s) {
		$zip = $1;
print "-------zip=$zip=\n";
	}
	print $zip;
	exit(0);
}

=cut

if (open(LZ, "$ENV{HOME}/localzip"))
{
	$zip = <LZ>;
	chomp $zip;
	close LZ;
	if ($zip =~ /\d/)
	{
		print $zip;
		exit(0);
	}
}
print '';
exit(1);	
